core/serialization/blocks.js

```diff
@@ -441,14 +441,17 @@ const loadAttributes = function(block, state) {
 /**
  * Applies any extra state information available on the state object to the
  * block.
+ * For json serialised blocks with a loadExtraState method, if the extraState
+ * is an xml mutation (not an object), domToMutation will be called instead for
+ * backward compatibility.
  * @param {!Block} block The block to set the extra state of.
  * @param {!State} state The state object to reference.
  */
 const loadExtraState = function(block, state) {
   if (!state['extraState']) {
     return;
   }
-  if (block.loadExtraState) {
+  if (block.loadExtraState && typeof state['extraState'] === 'object') {
```

Comment:
Could you move this check into the `loadExtraState` property of the `lists_getIndex` block instead?

Technically we don't require that the extraState be an object. So if someone has a custom block, where they've saved the state as a string, then this would break their implementation.

That does mean that if we have to do this upgrade for any other blocks, we will have to duplicate a bit of logic (which your current implementation avoids). But I still think moving this check is the better option for backwards compatibility reasons.

Reply:
I've considered some alternatives (block having a method that says whether to use loadExtraState or domToMutation, throwing an exception in loadExtraState that triggers domToMutation being used).

None of them seem worth it, so I'll cope with some duplicated logic :)